src/render/pipelines/std-rigged.ts

```diff
@@ -0,0 +1,292 @@
+import { EM } from "../../ecs/entity-manager.js";
+import { vec2, vec3, vec4, quat, mat4, V } from "../../matrix/sprig-matrix.js";
+import { assert, assertDbg } from "../../utils/util.js";
+import { computeTriangleNormal } from "../../utils/utils-3d.js";
+import { randColor } from "../../utils/utils-game.js";
+import { CY, CyMeshPoolPtr } from "../gpu-registry.js";
+import { createCyStruct, CyToTS } from "../gpu-struct.js";
+import { createMeshPool, MAX_INDICES, MeshHandle } from "../mesh-pool.js";
+import {
+  getAABBFromMesh,
+  isRigged,
+  Mesh,
+  RiggedMesh,
+} from "../../meshes/mesh.js";
+import {
+  mainDepthTex,
+  MeshUniformStruct,
+  MeshUniformTS,
+  positionsTexturePtr,
+  RenderDataStdDef,
+  sceneBufPtr,
+  surfacesTexturePtr,
+  unlitTexturePtr,
+  worldNormsAndFresTexPtr,
+} from "./std-scene.js";
+import { CyResources } from "../instantiator-webgpu.js";
+import { createCyArray } from "../data-webgpu.js";
+import { GPUBufferUsage } from "../webgpu-hacks.js";
+import { Renderer } from "../renderer-ecs.js";
+
+const MAX_MESHES = 20000;
+const MAX_VERTICES = MAX_INDICES; // 21844;
+// uniform buffer size is limited to 64k
+const MAX_JOINTS = 200;
+
+interface RiggedMeshWithJointIdx extends RiggedMesh {
+  jointIdx: number;
+}
+
+export interface RiggedMeshHandle extends MeshHandle {
+  jointIdx: number;
+}
+
+function isRiggedMeshWithJointIdx(m: Mesh): m is RiggedMeshWithJointIdx {
+  return isRigged(m) && "jointIdx" in m;
+}
+
+const VertexStruct = createCyStruct(
+  {
+    // TODO: this could be vec4<u8>
+    jointIds: "vec4<u32>",
+    jointWeights: "vec4<f32>",
+    position: "vec3<f32>",
+    color: "vec3<f32>",
+    normal: "vec3<f32>",
+    // TODO(@darzu): add UV back? needed for ocean stuff?
+    uv: "vec2<f32>",
+    surfaceId: "u32",
+  },
+  {
+    isCompact: true,
+    serializer: (
+      { jointIds, jointWeights, position, color, normal, uv, surfaceId },
+      _,
+      offsets_32,
+      views
+    ) => {
+      views.u32.set(jointIds, offsets_32[0]);
+      views.f32.set(jointWeights, offsets_32[1]);
+      views.f32.set(position, offsets_32[2]);
+      views.f32.set(color, offsets_32[3]);
+      views.f32.set(normal, offsets_32[4]);
+      views.f32.set(uv, offsets_32[5]);
+      views.u32[offsets_32[6]] = surfaceId;
+    },
+  }
+);
+type VertexTS = CyToTS<typeof VertexStruct.desc>;
+function createEmptyVertexTS(): VertexTS {
+  return {
+    jointIds: vec4.create(),
+    jointWeights: vec4.create(),
+    position: vec3.create(),
+    color: vec3.create(),
+    // tangent: m.tangents ? m.tangents[i] : [1.0, 0.0, 0.0],
+    normal: vec3.create(),
+    uv: V(0, 0),
+    surfaceId: 0,
+  };
+}
+
+const JointStruct = createCyStruct(
+  {
+    jointMatrix: "mat4x4<f32>",
+  },
+  {
+    isUniform: true,
+    // TODO: daryl, why is this necessary? it fixes things :(
+    hackArray: true,
```

Comment:
yeah.. i need to do another cytochrome pass around alignment and sizes. There's a few things like this that have built up.